request.go

```diff
@@ -0,0 +1,153 @@
+	"github.com/notaryproject/tspclient-go/internal/oid"
+)
+
+// MessageImprint contains the hash of the datum to be time-stamped.
+//
+//	MessageImprint ::= SEQUENCE {
+//	 hashAlgorithm   AlgorithmIdentifier,
+//	 hashedMessage   OCTET STRING }
+type MessageImprint struct {
+	HashAlgorithm pkix.AlgorithmIdentifier
+	HashedMessage []byte
+}
+
+// Request is a time-stamping request.
+//
+//	TimeStampReq ::= SEQUENCE {
+//	 version         INTEGER                 { v1(1) },
+//	 messageImprint  MessageImprint,
+//	 reqPolicy       TSAPolicyID              OPTIONAL,
+//	 nonce           INTEGER                  OPTIONAL,
+//	 certReq         BOOLEAN                  DEFAULT FALSE,
+//	 extensions      [0] IMPLICIT Extensions  OPTIONAL }
+type Request struct {
+	Version        int // fixed to 1 as defined in RFC 3161 2.4.1 Request Format
+	MessageImprint MessageImprint
+	ReqPolicy      asn1.ObjectIdentifier `asn1:"optional"`
+	Nonce          *big.Int              `asn1:"optional"`
+	CertReq        bool                  `asn1:"optional,default:false"`
+	Extensions     []pkix.Extension      `asn1:"optional,tag:0"`
+}
+
+// RequestOptions provides options for caller to create a new timestamp request
+type RequestOptions struct {
+	// Content is the datum to be time stamped. REQUIRED.
+	Content []byte
+
+	// HashAlgorithm is the hash algorithm to be used to hash the Content.
+	// REQUIRED and MUST be an available hash algorithm.
+	HashAlgorithm crypto.Hash
+
+	// HashAlgorithmParameters is the parameters for the HashAlgorithm.
+	// OPTIONAL.
+	HashAlgorithmParameters asn1.RawValue
+
+	// ReqPolicy specifies the TSA policy ID. OPTIONAL.
+	ReqPolicy asn1.ObjectIdentifier
+
+	// Nonce is a large random number with a high probability that the client
+	// generates it only once. The same nonce is included and validated in the
+	// response. OPTIONAL.
+	Nonce *big.Int
+
+	// CertReq determines if TSA signing certificate is included in the response.
+	// OPTIONAL.
+	CertReq bool
+
+	// Extensions is a generic way to add additional information
+	// to the request in the future. OPTIONAL.
+	Extensions []pkix.Extension
+}
+
+// NewRequest creates a timestamp request based on caller provided options.
+func NewRequest(opts RequestOptions) (*Request, error) {
+	if opts.Content == nil {
+		return nil, &MalformedRequestError{Msg: "content to be time stamped cannot be empty"}
+	}
+	hashAlg, err := oid.FromHash(opts.HashAlgorithm)
+	if err != nil {
+		return nil, &MalformedRequestError{Msg: err.Error()}
+	}
+	digest, err := hashutil.ComputeHash(opts.HashAlgorithm, opts.Content)
+	if err != nil {
+		return nil, &MalformedRequestError{Msg: err.Error()}
+	}
+	return &Request{
+		Version: 1,
+		MessageImprint: MessageImprint{
+			HashAlgorithm: pkix.AlgorithmIdentifier{
+				Algorithm:  hashAlg,
+				Parameters: opts.HashAlgorithmParameters,
+			},
+			HashedMessage: digest,
+		},
+		ReqPolicy:  opts.ReqPolicy,
+		Nonce:      opts.Nonce,
+		CertReq:    opts.CertReq,
```

Comment:
IMO defualt value of CertReq should be true.

Reply:
According to [RFC 3161 2.4.1](https://datatracker.ietf.org/doc/html/rfc3161#section-2.4.1), `certreq` is `DEFAULT FALSE`.

Reply:
Yes but we should optimize for more common usecase

Reply:
Then user would surprise that the library is not RFC compliant.

Reply:
In most of the usecases user will want cert. If user doesn't want cert, they can explicitly set it to false. IMO RFC is more for TSA(server) than client.

Reply:
Makes sense. Updating the default value to `true`. We could discuss this further if other reviewers have different opinions.

Reply:
Updated in the struct `RequestOptions`, as this is the struct that users will use to create a `Request` through constructor `NewRequest` with this library.
This is a good place to answer your previous question regarding `RequestOptions` and `Request`: 
1. The purpose for `RequestOptions` is to make the code extensible. When a new field is introduced, adding it into `RequestOptions` is desirable comparing to modify the parameters of `NewRequest`, which is a breaking change. This pattern can be found in many Go projects and other Notary Project repos such as notation-go.
2. We need `Request` to be exported, because the output of constructor `NewRequest` is `Request`. The fields under `Request` are exported as well due to implementation constraints of `asn1.Unmarshal`, if we do not expose all the fields in `Request`, we will get `unexported field` error during unmarshalling.
3. `RequestOptions` can have different fields than `Request`. `Request` is essentially the struct defined in RFC 3161, for general purpose, it is, by design, strictly following its definition in the RFC. However, `RequestOptions` allows users of this library to pick their own configuration. For example, `NoCert` (default as false) changes the `CertReq` field default to `true` in the `NewRequest` constructor.

Reply:
Resolving as discussed in 6/11/24 community meeting.